afqinsight/cross_validate.py

```diff
@@ -464,3 +469,71 @@ def cross_validate_checkpoint(
             ret[key] = np.array(train_scores[name])
 
     return ret
+
+
+def stratified_crossvalidate(
+    covariate_df,
+    pipeline_estimator=None,
+    cv_splitter=StratifiedShuffleSplit(n_splits=1, test_size=0.25),
+):
+    """
+    Given multiple covariates in covariate_df, returns training and test splits
+    such that covariate distributions are similar across both training and test splits
+
+    Parameters
+    ----------
+    covariate_df : pandas dataframe of shape (n_samples, n_features)
+
+    pipeline_estimator : default=None
+        Custom pipeline not yet supported
+
+    cv_splitter : A cv object default=StratifiedShuffleSplit
+        A scikit-learn object for splitting the dataset into
+        train/test set such that splits are stratified
+
+    Yields
+    ------
+    train : ndarray
+        The training set indices for one shuffle split
+    test : ndarray
+        The testing set indices for one shuffle split
+    covariate: ndarray
+        The composite covariate used to stratify training and test splits
+
+    """
+
+    # Check covariate_df is a dataframe
+    if not isinstance(covariate_df, pd.DataFrame):
+        raise TypeError
```

Comment:
```suggestion
        raise TypeError("`covariate_df` input needs to be a DataFrame object")
```

Is that crucial, by the way? Why not take arrays?